### PR TITLE
[Model 8] REST API polishing, cleanup and extensions

### PR DESCRIPTION
**Pull request by afein**
This PR is focused in 2 major changes to the model REST API:
- Model GetXMetric methods now receive specific Request structs as parameters.
- Added new Getters,  endpoints and tests for extracting all nodes, namespaces, pods and containers from the model. This should allow API consumers to seamlessly 

\cc @rjnagal @vmarmol

### CODE REVIEW

File: api/v1/model_handlers.go [hunk 1]
```diff
@@ -35,12 +36,19 @@ func (a *Api) RegisterModel(container *restful.Container) {
...
+		Doc("Get a list of all entities available in the model").
+		Operation("allEntities"))
+
+	// The /cluster/ endpoint returns a list of all available metrics for the Cluster entity of the model.
+	ws.Route(ws.GET("/cluster/").
 		To(a.availableMetrics).
 		Filter(compressionFilter).
-		Doc("Show which metrics are available").
-		Operation("clusterMetrics"))
+		Doc("Get a list of all available metrics for the Cluster entity.").
```
Comment (mvdan) [inline api/v1/model_handlers.go]: Period should probably go.

File: api/v1/model_handlers.go [hunk 2]
```diff
@@ -50,8 +58,25 @@ func (a *Api) RegisterModel(container *restful.Container) {
...
+		Filter(compressionFilter).
+		Doc("Get a list of all Nodes in the model").
+		Operation("allNodes").
 		Writes(MetricResult{}))
 
+	// The /nodes/{node-name} endpoint returns a list of all available metrics for a Node entity.
+	ws.Route(ws.GET("/nodes/{node-name}").
+		To(a.availableMetrics).
+		Filter(compressionFilter).
+		Doc("Get a list of all available metrics for a Node entity.").
```
Comment (mvdan) [inline api/v1/model_handlers.go]: Same here.

File: api/v1/model_handlers.go [hunk 3]
```diff
@@ -267,11 +388,11 @@ func (a *Api) freeContainerMetrics(request *restful.Request, response *restful.R
 	response.WriteEntity(exportTimeseries(timeseries, new_stamp))
 }
 
-// parseRequestStartParam obtains the time.Time to be used as the start time during a model Get method.
-// parseRequestStartParam receives a request and a response as inputs, and returns the parsed time.
-func parseRequestStartParam(request *restful.Request, response *restful.Response) time.Time {
+// parseRequestParam parses a time.Time from a named QueryParam.
+// parseRequestParam receives a request and a response as inputs, and returns the parsed time.
+func parseRequestParam(param string, request *restful.Request, response *restful.Response) time.Time {
```
Comment (mvdan) [inline api/v1/model_handlers.go]: nit: shouldn't extra arguments be at the end? Also, looks like its doc needs updating.

File: model/impl.go [hunk 4]
```diff
@@ -161,54 +160,136 @@ func (rc *realCluster) GetPodContainerMetric(namespace string, pod_name string,
...
+	if !ok {
+		return res
+	}
+
+	podref, ok := ns.Pods[pod]
+	if !ok {
+		return res
+	}
+
+	for key, _ := range podref.Containers {
```
Comment (mvdan) [inline model/impl.go]: nit: `, _` can be removed

**Comment by afein** [comment]
updated the API to provide metrics under /metrics for each entity, thus enhancing browsability of the API and allowing a future /stats extension to be added to each entity.
Also addressed most of mvdan's comments.

File: api/v1/model_handlers.go (hunk 3 above)
Comment (afein) [inline api/v1/model_handlers.go]: I think that parseRequestParam("start", request, response) is more readable than parseRequestParam(request, response, "start"), especially for consecutive calls with the same request/response arguments.

**Comment by mvdan** [comment]
retest this please

**Comment by mvdan** [comment]
Jenkins is trying this on kubernetes 0.20.0, this is weird.

**Pull request closed without merging**